### ISSUE TITLE
Semantic Versioning Policy

### ISSUE THREAD

**Issue by kjin** [google/gts#63]
@ofrobots brought this up in a discussion a few days ago on how we should go forward in following semver guidelines for `gts`, with the central issue that Google's TS style guide is a moving target, adopting new changes over time. To ensure that users of `gts` don't get caught unaware, it's important for us to consider how we can map changes to the style guide to semantic versioning levels in a way that makes sense.

**Comment by JustinBeckwith** [comment]
Version numbers are cheap, and I think it's important to let developers know what they're getting into with a minor version update.  If possible, I really think we should do a semver major change if we add a new rule, or change anything that could be considered a "build break".  Would that be a reasonable approach?

**Comment by ofrobots** [comment]
Version numbers are cheap, but I think it is high burden on the users of this module if breaking changes arrive very frequently and require upkeep of their codebase on a regular basis. I think it might incentivize people to lock down to a specific versions.

A counter perspective here is that we do provide `gts fix` that can fix most (but not all) problems. An argument could be made that we _could_ get away with not considering auto-fixable changes as semver major. However, a frequent occurrence of these would still be problematic as it requires actual work (even if it is cheap) to stay up to date.

I propose that we limit to N (say 2) semver major releases a year so that the cadence is not too frequent. In other words we may lag behind the 'official' google style guide by some number of months until we do a semver major release with the updates.

I am very interested in other opinions on this too.

**Comment by JustinBeckwith** [comment]
I've come around to your point of view here, especially since it's consistent with what TypeScript and a few other tools do.  

If we're in agreement here @ofrobots, I'd be happy to submit a README PR letting users know the policy.

**Comment by JustinBeckwith** [comment]
@google/google-node-team speak now if you're against breaking changes in minor versions :)

**Comment by ofrobots** [comment]
I presented both sides of the arguments in my previous comment. I didn't specify a preference just yet – I do have one, but I don't want to bias the discussion.

**Comment by JustinBeckwith** [comment]
😆 Sorry.  To clearly state my position:
- I think we should treat changes in linting and formatting as semver *minor* changes.  This is consistent with TypeScript (for better or worse).
- If we make changes breaking changes to the API, I'm in favor of a semver major.
- We should limit releases that have new versions of tslint / clang-format to reduce churn.  I'd be ok with once a quarter, though every six months may work too.

**Comment by alexander-fenster** [comment]
Speaking from the user perspective here, I remember that prettier introduced a breaking change (by fixing a bug or in some other way) in 1.11 which broke the CI for our common package and forced me to update the code (https://github.com/googleapis/nodejs-common/pull/47). That bug made me spend some time understanding what has just happened and why the CI went mad, and I'd rather spend this time on something more useful.

In other words - I vote for not introducing incompatibilities in minor versions.

**Comment by ofrobots** [comment]
Yes, it is incredibly annoying to see your CI all red after a README.md change, just because *some* tool, whether it is clang-format (which doesn't follow semver) or typescript (which doesn't follow semver), makes a breaking change that you just picked up. It doesn't matter if automatic fixers are available; your CI is all red, and the change you wanted to make is blocked because you have to figure out who who is mad and why. Not all our users use package-locks. Even the users that use package-locks would be incentivized to not accept their greenkeeper PRs.

I strongly believe that we must follow semver – that is what JS developers expect. I think the typescript project got this wrong.

I would go further to say that releasing semver major changes too frequently is too much pain. For this project, I think we should do no more than 2 semver majors in a year.

**Comment by JustinBeckwith** [comment]
That sounds good too :) I just wanted to see us all come to some sort of conclusion here.  So it sounds like we're saying:
- Any changes to formatting or lint rules will be treated as semver major
- We will make two semver major releases a year

If we're all on the same page, I will go update the README and make this a policy.  Thoughts?

**Comment by ofrobots** [comment]
> Any changes to formatting or lint rules will be treated as semver major
Breaking changes are semver major. So if I remove a lint rule, that is not major. 

> We will make two semver major releases a year
Let's say we will target around 2 semver major releases a year. Let's leave some flexibility in there.

**Comment by jinwoo** [comment]
> So if I remove a lint rule, that is not major.
This is debatable. What if the client repo had a test that asserts a failure for that specific lint rule? That's very unlikely but still...

**Comment by ofrobots** [comment]
I would say that lies outside our contract with the users. It is reasonable for the users to expect that a semver patch or minor update will not cause `gts check` or compile failures. We do not have a contract on the precise list of lint rules we run.

**Comment by ofrobots** [comment]
We follow semver
